### ISSUE TITLE
EasyRdf_Resource::allResources() does not work with fully qualified URIs

### ISSUE THREAD

**Issue by zozlak** [njh/easyrdf#275]
When calling the `allResources()` method on the `EasyRdf_Resource` object you must pass the property URI in a prefixed form because the fully qualified form is not supported.

Surprisingly some other methods (e.g. `hasProperty()`) support both prefixed and fully qualified forms.

Code example below - we would like to get `1 1 1` but we get `1 0 1`.

Tested on EasyRdf 0.9.1

```php
$property = 'http://purl.org/dc/terms/identifier';
$graph = new EasyRdf_Graph();
$resource = $graph->resource('https://my.domain/my/uri');
$resource->addResource($property, 'http://my.domain/sample#id');
// the property is there
echo $resource->hasProperty($property) . "\n";
// can't be fetched using the fully qualified URI
echo count($resource->allResources($property)) . "\n";
// but works when written in a prefixed form
$prefixed = join(':', EasyRdf_Namespace::splitUri($property, true));
echo count($resource->allResources($prefixed)) . "\n";
```

**Comment by njh** [comment]
Just going through some old issues and assigning labels to them.
Sorry, this issue is 4 years old 😢 

It isn't good that the different methods behave in different ways.
This is partly because some methods accept property paths - like SPARQL.
Others only take a single/specific property.
But maybe this needs some clarification.

So there are a couple of ways round this.

First, use `<` `>` round URIs:
```php
echo count($resource->allResources("<$property>")) . "\n";
```

Alternatively use a resource:
```php
$property = $graph->resource('http://purl.org/dc/terms/identifier');
echo count($resource->allResources($property)) . "\n";
```

**Comment by zozlak** [comment]
Yep, I'm aware of the workarounds (was able to find them during four years :smile:  ), just I find the current behaviour so highly undesirable that I won't go back to EasyRdf (from my own fork) until the API is made orthogonal.

I dislike the current behaviour for a few reasons:

* It makes the API non-orthogonal (different methods behave in a different way with the same parameter denoting the same thing).
    * Which makes it more difficult to learn and use (despite some specific cases which might had been crucial for a person who implemented it). The example provided at the top of this issue is really confusing for a new user (especially taking into account gaps in the documentation described below).
* It's not inline with the API documentation. All `Resource` class methods taking the property-like parameter describe it as *$property The name of the property (e.g. foaf:name)* while in practice it's interpreted as *a property or a SPARQL path like expression*.
* The SPARQL path like expressions are far from real SPARQL path expressions. The parser is extremely simple (just splits on a slash and for each term splits it on a pipe and finally for each resulting terms checks if it begins with a hat) and will break on many completely valid SPARQL expressions (e.g. ones using fully qualified URIs :smiley: , ones with brackets, ones using ?, + or * qualifiers).
    * Moreover the documentation doesn't say a word about the subset of SPARQL path expressions being supported. In the documentation we only see *@param  string  $propertyPath  A valid property path*

All in all the SPARQL-like feature looks to me like something added *quickly & dirty* because it simplified someone's pretty particular workflow at the time of adding it.

In my opinion the right implementation would be (sorted from less to more severe API changes):

* either to make all methods handle the SPARQL path like expressions (with e.g. `Resource::hasProperty($prop)` from the example above either returning null or throwing an error) 
* or to make a SPARQL path like expressions a separate class being passed to `Graph` and `Resource` methods (so every method can easily distinguish if a parameter is *just a property name* or *a SPARQL path like expression*) just like now you advice to pass the property URI as a `Resouce` object instead of a string
* or to implement the SPARQL path like expressions as a higher software layer by creating a *SPARQL path expressions executor class* which would take an expression and a graph and execute it on the graph (which probably provides the best software layers separation, simplicity of API and the code and testability of the whole codebase)

No matter the solution, the documentation should be made clear (also on the supported subset of the SPARQL path expressions).

To make it clear - I'm not requesting any immidiate change in the EasyRdf here. I have my own fork which I adjusted to my needs and I can keep using it. On the other hand with the recent discussion on the EasyRdf future I think it's a good time to bring up and discuss some EasyRdf features which I consider not so well implemented and which (in my opinion) could be improved (also with my help).

**Comment by k00ni** [comment]
Your arguments make sense to me. 

As a next step I suggest we adapt the API doc so that it reflects the current state. That shouldn't be that time consuming. 

 @njh and @zozlak if you are good with that, what method headers need to be changed?

**Comment by k00ni** [comment]
Ping

**Comment by zozlak** [comment]
It affects all getters (`getResource()`, `getLiteral()`, `get()`, `allResources()`, `allLiterals()` and `all()`) both on `Resource` and `Graph` class. The `__get()` magic on `Resource` is affected as well.

```php
$g = new EasyRdf\Graph();
$u = 'http://foo#bar';
$r = $g->resource($u);
$p = 'http://foo#baz';
$r->addResource($p, 'http://bar#baz');
$r->addLiteral($p, 'baz');

var_dump($r->allResources($p));
var_dump($r->allLiterals($p));
var_dump($r->all($p));
var_dump($r->get($p));
var_dump($r->getResource($p));
var_dump($r->getLiteral($p));
var_dump($r->{$p});
var_dump($g->allResources($u, $p));
var_dump($g->allLiterals($u, $p));
var_dump($g->all($u, $p));
var_dump($g->get($u, $p));
var_dump($g->getResource($u, $p));
var_dump($g->getLiteral($u, $p));

$pp = join(':', EasyRdf\RdfNamespace::splitUri($p, true));
var_dump($r->allResources($pp));
var_dump($r->allLiterals($pp));
var_dump($r->all($pp));
var_dump($r->get($pp));
var_dump($r->getResource($pp));
var_dump($r->getLiteral($pp));
var_dump($r->{$pp});
var_dump($g->allResources($u, $pp));
var_dump($g->allLiterals($u, $pp));
var_dump($g->all($u, $pp));
var_dump($g->get($u, $pp));
var_dump($g->getResource($u, $pp));
var_dump($g->getLiteral($u, $pp));
```

**Comment by k00ni** [comment]
Are you OK with adapting the doc first and hopefully deal with the code later?

**Comment by zozlak** [comment]
I think it's the only option.

**Comment by Michiel-s** [comment]
Let me start by complimenting @njh and the EasyRdf development team with this nice library. Thank you!

The issue here took me quite some time to figure out and diagnose myself. Especially the misleading php doc comments mentioned above in combination with different behavior between getters and setters. Currently I'm using v1.1.1 discovered the issue myself for the `Graph::allLiterals` method.

Looking forward for a solution. In the meantime I'll use the workaround mentioned above using a Resource:
```
$property = $graph->resource('http://purl.org/dc/terms/identifier');
echo count($resource->allResources($property)) . "\n";
```

**Comment by Michiel-s** [comment]
Same issue reported in #128 for allLiterals() method only.
